GVFS/GVFS.Platform.Mac/MacFileSystem.cs

```diff
@@ -43,33 +46,71 @@ public bool HydrateFile(string fileName, byte[] buffer)
             return NativeFileReader.TryReadFirstByteOfFile(fileName, buffer);
         }
 
+        public unsafe void WriteFile(ITracer tracer, byte* originalData, long originalSize, string destination)
+        {
+            int fileDescriptor = 1;
+            try
+            {
+                fileDescriptor = NativeFileReader.Open(destination, NativeFileReader.WriteOnly | NativeFileReader.Create);
+                IntPtr result = Write(fileDescriptor, originalData, (IntPtr)originalSize);
+                if (result.ToInt32() == -1)
+                {
+                    throw new Win32Exception(Marshal.GetLastWin32Error());
+                }
+            }
+            catch
+            {
+                tracer.RelatedError("Error writing file {0}. ERRNO: {1}", destination, Marshal.GetLastWin32Error());
+                throw;
+            }
+            finally
+            {
+                NativeFileReader.Close(fileDescriptor);
+            }
+        }
+
         [DllImport("libc", EntryPoint = "chmod", SetLastError = true)]
         private static extern int Chmod(string pathname, int mode);
 
         [DllImport("libc", EntryPoint = "rename", SetLastError = true)]
         private static extern int Rename(string oldPath, string newPath);
 
+        [DllImport("libc", EntryPoint = "write", SetLastError = true)]
+        private static unsafe extern IntPtr Write(int fileDescriptor, void* buf, IntPtr count);
+
         private class NativeFileReader

```

Comment:
I debated restructuring this class here since it's no longer just being used for the FileReader code in Prefetch, but I think it might make sense as a follow up PR so this one doesn't get too large.